### ISSUE TITLE
Improvements to public API documentation

### ISSUE THREAD

**Issue by tonybaloney** [kubernetes/kubernetes#19680]
the public API documentation on kubernetes.io is very challenging to understand compared with projects like Docker, OpenStack.

The first and most challenging issue is that all of the API operations
are listed on one massive page, there isn't any grouping or index. I
have to use the browser find to search within the page and guess what
I'm looking for.

The second issue is that the docs are automatically generated so a lot
of the descriptions are not helpful, for example "proxy PUT requests
to Pod" : PUT /api/v1/proxy/namespaces/{namespace}/pods/{name} - I
have no idea what this does or what it is supposed to do.

Third issue is that the models 'required' for each of the fields are
not entirely correct. I was trying to create a pod. The docs say :
http://kubernetes.io/v1.1/docs/api-reference/v1/definitions.html#_v1_pod
that a pod spec is required, but none of the fields are required.
The API returns and gives an error saying fields are missing. I had to
use trial and error to discover which fields are required at a minimum
(turns out, metadata.name, containers[0].name and
containers[0].image).

@caesarxuchao
@brendandburns
@bgrant0607

**Comment by caesarxuchao** [comment]
Thanks @tonybaloney. I'll send fixes next week.

**Comment by bgrant0607** [comment]
Related: #14924, #11933, #16095

The problem with PodSpec is that the same type is shared between Pod and PodTemplate.

**Comment by caesarxuchao** [comment]
> The problem with PodSpec is that the same type is shared between Pod and PodTemplate.

@bgrant0607, I don't think this affects if the fields of `PodSpec` are `required`, because the validation functions of `Pod` and `PodTemlate` both call the same function to validate PodSpec: [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L1310) and [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L1679).

What's our convention for fields that "one of the fields needs to be specified", like this [one](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L2163)? Shall we mark both as not required?

**Comment by thockin** [comment]
We have to mark them both as optional.  AFAIK we have no real way to denote
one-of.  Maybe we need a new field tag like 'oneof:foobar' and only one
field tagged as such must be set.

**Comment by bgrant0607** [comment]
Also ref: #17404, #17362, #16873, #16550, #16095, #16049, #15910, #15741, #15244, #15128, #14924, #14189, #11933, #11567, #11468, #11332, #11000, #10710, #10506, #8528, #6167, #6145, #4934, #4931, #4700, #3714, #3606, #2971, #1963

**Comment by nikhiljindal** [comment]
Another problem is that we use the same Pod model for different operations like create and update and the validations for both are different.

**Comment by bgrant0607** [comment]
Also #22389

We've reached the point where we desperately need good API documentation.

cc @lavalamp

**Comment by bgrant0607** [comment]
Need to replace out-of-date design docs and proposals: #22797

**Comment by bgrant0607** [comment]
Example API documentation needed: content type: #22904

**Comment by bgrant0607** [comment]
Doc overhaul in progress:
https://docs.google.com/presentation/d/14mJwJGlwBJHIpMRGR-nYMSyyNk-kRpabyz74-YYXeTo/edit#slide=id.g12d21f53ed_2_11

It won't specifically address many of these issues, but hopefully will provide a better framework in which to address them.

**Comment by bgrant0607** [comment]
https://github.com/kubernetes/kubernetes.github.io/issues/650

**Comment by bgrant0607** [comment]
Another issue: example value for each field

**Comment by timbunce** [comment]
The [Elasticsearch API documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/index.html) is worth a look as an example.

I especially appreciate that it's easy to switch between versions for a given page, e.g. here's the [cluster page for version 1.5](https://www.elastic.co/guide/en/elasticsearch/reference/1.5/cluster.html). At the top of the page is a clear message saying it's an old version and giving a link to the current. On the right is a drop-down making it trivial to switch to other versions of those docs.

**Comment by bgrant0607** [comment]
I don't remember whether it is specifically mentioned somewhere else, but we need a developer-facing version of:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md

**Comment by zhangxiaoyu-zidif** [comment]
@bgrant0607 
Your https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md
return 404 error now

**Comment by zhangxiaoyu-zidif** [comment]
I found it in https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md
:)

**Comment by sftim** [comment]
Following https://github.com/kubernetes/kubernetes/issues/19680#issuecomment-237964227 I've logged issue #89094

**Comment by tonybaloney** [comment]
@thockin this issue is getting a little old :-) Should I close it?

**Comment by sftim** [comment]
I think this issue is worth keeping. However, @tonybaloney, perhaps you'd be willing to update the issue description to account for the changes / improvements / regressions since you opened it?

**Comment by thockin** [comment]
I think it's still a real issues

**Comment by Abhijais4896** [comment]
Can I get assigned to this?

**Comment by thockin** [comment]
Declarative Validation is a new source of rich information which can be used to generate better docs.  

@lalitc375 @yongruilin @aaron-prindle
